### ISSUE TITLE
VAO and dynamic Geometry change crash

### ISSUE THREAD

**Issue by valid-ptr** [openscenegraph/OpenSceneGraph#407]
There is a crash in Geometry while adding new array (texture or any other).

Try to run this example:

```
// NOTE: VAO "dynamic geometry test"
// g++ vao_dynamic_geom.cpp -I /opt/OpenSceneGraph/include/ -L /opt/OpenSceneGraph/lib64/ -losgViewer -losg -losgDB -losgGA
// Run with and without OSG_VERTEX_BUFFER_HINT=VERTEX_ARRAY_OBJECT / VERTEX_BUFFER_OBJECT (under gdb :)
// Example: OSG_WINDOW="0 0 800 450" OSG_VERTEX_BUFFER_HINT=VERTEX_ARRAY_OBJECT ./a.out
// Press "c" to crash app!

#include <iostream>
#include <osgViewer/Viewer>

struct Crash : public osgGA::GUIEventHandler
{
	osg::ref_ptr<osg::Geometry>		_Geometry;

	Crash(osg::Geometry* aGeometry)
	:	_Geometry(aGeometry)
	{
	}

	virtual bool handle(const osgGA::GUIEventAdapter& ea, osgGA::GUIActionAdapter&)
	{
		if (ea.getKey() == osgGA::GUIEventAdapter::KEY_C)	// push, release, doesn't matter
			// Crash
		{
			std::cout << "Add texture coords 1...." << std::endl;

			osg::Vec2Array* tcoords = new osg::Vec2Array(4);
			(*tcoords)[0].set(0.0f,0.0f);
			(*tcoords)[1].set(1.0f,0.0f);
			(*tcoords)[2].set(1.0f,1.0f);
			(*tcoords)[3].set(0.0f,1.0f);
			_Geometry->setTexCoordArray(1, tcoords);		// CRASH with VAO; number 1,2,3 etc (not 0 which exists on geometry)

			std::cout << "Texture coords 1 added!" << std::endl;
		}

		return false;
	}
};

int main(int argc, char** argv)
{
	osg::ArgumentParser parser(&argc, argv);
	osgViewer::Viewer viewer(parser);

	osg::ref_ptr<osg::Geometry> geometry =
		osg::createTexturedQuadGeometry(osg::Vec3d(-0.5, 0.0, -0.5),
										osg::Vec3d(1.0, 0.0, 0.0),
										osg::Vec3d(0.0, 0.0, 1.0));
	geometry->setDataVariance(osg::Object::DYNAMIC);


	osg::ref_ptr<osg::Group> root = new osg::Group;
	root->addChild(geometry);

	viewer.addEventHandler(new Crash(geometry));
	viewer.setSceneData(root);
	viewer.run();

	return 0;
}
```

**Comment by openscenegraph** [comment]
Thanks for the test, using your suggestions I have been able to reproduce to the crash.   I will now look into the cause and potential fixes

**Comment by openscenegraph** [comment]
I have tracked down the bug to the VertexArrayState object not being reinitialized to account for the new texture coord arrays.  I haven't yet resolved how best to force a reinitialization of the VertexArrayState.

**Comment by openscenegraph** [comment]
I have resolved the bug by improving the way that the VertexArrayState object is maintained so that new vertex array usage gets properly handled.  

https://github.com/openscenegraph/OpenSceneGraph/commit/5afd32b2d9960eb7b686f5b25b50e16f157c84d8

Quite a bit of work to resolve a bug, but the bug was really a missing feature of the original design & implementation, so required completing it.  Thanks for the test case this really helped in pinpointing the problem and providing a test case to make sure it was resolved.
